File: src/attestations/envelope.zig
// SPDX-License-Identifier: MIT OR Apache-2.0
//
// DSSE envelope — the outer signed container for every mkit attestation.
// See docs/SPEC-ATTESTATIONS.md §2 and
// https://github.com/secure-systems-lab/dsse/blob/master/envelope.md
//
// Envelope shape (JCS-canonical JSON):
//   {
//     "payload":     "<base64(payload_bytes)>",
//     "payloadType": "<media-type>",
//     "signatures":  [ { "keyid": <string>, "sig": "<base64(sig_bytes)>" }, ... ]
//   }
//
// Signed bytes are the DSSE Pre-Authentication Encoding (PAE):
//   "DSSEv1" SP ascii(len(type)) SP type SP ascii(len(payload)) SP payload
// where `payload` is the UTF-8 bytes (not base64) of the payload.
//
// mkit does not implement the hashed-PAE variant (DSSEv1 + SHA-256); we
// stick to the one-variant-only form because our signer trait always
// hands the signer the raw PAE and lets it decide whether to pre-hash.

const std = @import("std");
const Allocator = std.mem.Allocator;

const jcs = @import("jcs.zig");
const hash_mod = @import("../hash.zig");
const Hash = hash_mod.Hash;

/// MIME type written into `payloadType` for every mkit attestation.
pub const PAYLOAD_TYPE_IN_TOTO = "application/vnd.in-toto+json";

pub const Signature = struct {
    keyid: []const u8,
    /// Raw signature bytes (not base64). Encoder base64s them on the way out.
    sig: []const u8,
};

pub const Envelope = struct {
    payload_type: []const u8,
    /// Raw payload bytes (not base64). Encoder base64s them on the way out.
    payload: []const u8,
    signatures: []const Signature,
};

/// Build the DSSE PAE for a `(payloadType, payload)` pair.
/// Returns allocator-owned bytes.
pub fn pae(allocator: Allocator, payload_type: []const u8, payload: []const u8) ![]u8 {
    return std.fmt.allocPrint(
        allocator,
        "DSSEv1 {d} {s} {d} {s}",
        .{ payload_type.len, payload_type, payload.len, payload },
    );
}

/// Encode an envelope to JCS-canonical JSON. Caller owns the returned slice.
pub fn encode(allocator: Allocator, env: Envelope) ![]u8 {
    if (env.signatures.len == 0) return error.EnvelopeNeedsAtLeastOneSignature;
    if (env.payload_type.len == 0) return error.PayloadTypeEmpty;

    // Pre-compute base64 buffers so we can stick them in a jcs.Value tree.
    const b64 = std.base64.standard.Encoder;
    const payload_b64 = try allocator.alloc(u8, b64.calcSize(env.payload.len));
    defer allocator.free(payload_b64);
    _ = b64.encode(payload_b64, env.payload);

    // signatures[] — allocate one jcs.Member list per signature. We track
    // per-entry base64 allocations so we can free them all on the way out.
    const sig_b64s = try allocator.alloc([]u8, env.signatures.len);
    defer {
        for (sig_b64s) |s| allocator.free(s);
        allocator.free(sig_b64s);
    }

    const sig_values = try allocator.alloc(jcs.Value, env.signatures.len);
    defer allocator.free(sig_values);

    const sig_members_storage = try allocator.alloc([2]jcs.Member, env.signatures.len);
    defer allocator.free(sig_members_storage);

    for (env.signatures, 0..) |s, i| {
        const buf = try allocator.alloc(u8, b64.calcSize(s.sig.len));
        _ = b64.encode(buf, s.sig);
        sig_b64s[i] = buf;

        sig_members_storage[i] = [2]jcs.Member{
            .{ .key = "keyid", .value = .{ .string = s.keyid } },
            .{ .key = "sig", .value = .{ .string = buf } },
        };
        sig_values[i] = .{ .object = sig_members_storage[i][0..] };
    }

    const root: jcs.Value = .{ .object = &.{
        .{ .key = "payload", .value = .{ .string = payload_b64 } },
        .{ .key = "payloadType", .value = .{ .string = env.payload_type } },
        .{ .key = "signatures", .value = .{ .array = sig_values } },
    } };

    return try jcs.encode(allocator, root);
}

/// Compute the attestation id: BLAKE3 over the encoded envelope bytes.
pub fn attestationId(envelope_bytes: []const u8) Hash {
    return hash_mod.hash(envelope_bytes);
}

// -----------------------------------------------------------------------------
// Decoder — minimal, only what `mkit attest verify` / `show` need.
//
// We do NOT round-trip arbitrary JSON. We accept the shape our encoder
// produces and extract the four fields we care about. Anything else is
// rejected with `error.MalformedEnvelope`. This is a trade-off — it keeps
// us from needing a full JSON parser, at the cost of being intolerant of
// attestations produced by other tools with non-JCS spacing.
//
// Mitigation: if we ever need to ingest third-party DSSE envelopes we
// re-canonicalise via their JSON + our writer before storing; on disk
// mkit only ever sees its own output, where the decoder's strictness
// is a feature.
// -----------------------------------------------------------------------------

pub const DecodedEnvelope = struct {
    payload_type: []u8,
    payload: []u8,
    signatures: []DecodedSignature,
    allocator: Allocator,

    pub fn deinit(self: *DecodedEnvelope) void {
        for (self.signatures) |s| {
            self.allocator.free(s.keyid);
            self.allocator.free(s.sig);
        }
        self.allocator.free(self.signatures);
        self.allocator.free(self.payload);
        self.allocator.free(self.payload_type);
    }
};

pub const DecodedSignature = struct {
    keyid: []u8,
    sig: []u8,
};

/// Decode the JCS-canonical DSSE envelope bytes our `encode` produces.
/// Accepts only the exact shape + key order we emit. Strict by design.
pub fn decode(allocator: Allocator, bytes: []const u8) !DecodedEnvelope {
    const b64 = std.base64.standard.Decoder;

    // Expected form:
    // {"payload":"<..>","payloadType":"<..>","signatures":[{"keyid":"..","sig":".."},...]}
    var p = Parser{ .src = bytes, .pos = 0 };
    try p.expect("{\"payload\":");
    const payload_b64 = try p.takeString(allocator);
    errdefer allocator.free(payload_b64);

    try p.expect(",\"payloadType\":");
    const payload_type = try p.takeString(allocator);
    errdefer allocator.free(payload_type);

    try p.expect(",\"signatures\":[");
    var sigs: std.ArrayList(DecodedSignature) = .empty;
    errdefer {
        for (sigs.items) |s| {
            allocator.free(s.keyid);
            allocator.free(s.sig);
        }
        sigs.deinit(allocator);
    }
    if (!p.peek(']')) {
        while (true) {
            try p.expect("{\"keyid\":");
            const keyid = try p.takeString(allocator);
            errdefer allocator.free(keyid);
            try p.expect(",\"sig\":");
            const sig_b64 = try p.takeString(allocator);
            errdefer allocator.free(sig_b64);
            try p.expect("}");

            const sig_bytes = try allocator.alloc(u8, b64.calcSizeForSlice(sig_b64) catch return error.MalformedEnvelope);
            errdefer allocator.free(sig_bytes);
            b64.decode(sig_bytes, sig_b64) catch return error.MalformedEnvelope;
            allocator.free(sig_b64);

            try sigs.append(allocator, .{ .keyid = keyid, .sig = sig_bytes });

            if (p.peek(',')) {
                p.pos += 1;
                continue;
            }
            break;
        }
    }
    try p.expect("]}");
    if (p.pos != bytes.len) return error.MalformedEnvelope;

    const payload = try allocator.alloc(u8, b64.calcSizeForSlice(payload_b64) catch return error.MalformedEnvelope);
    errdefer allocator.free(payload);
    b64.decode(payload, payload_b64) catch return error.MalformedEnvelope;
    allocator.free(payload_b64);

    return .{
        .payload_type = payload_type,
        .payload = payload,
        .signatures = try sigs.toOwnedSlice(allocator),
        .allocator = allocator,
    };
}

const Parser = struct {
    src: []const u8,
    pos: usize,

    fn expect(self: *Parser, s: []const u8) !void {
        if (self.pos + s.len > self.src.len) return error.MalformedEnvelope;
        if (!std.mem.eql(u8, self.src[self.pos .. self.pos + s.len], s)) return error.MalformedEnvelope;
        self.pos += s.len;
    }

    fn peek(self: *const Parser, c: u8) bool {
        return self.pos < self.src.len and self.src[self.pos] == c;
    }

    fn takeString(self: *Parser, allocator: Allocator) ![]u8 {
        if (self.pos >= self.src.len or self.src[self.pos] != '"') return error.MalformedEnvelope;
        self.pos += 1;
        const start = self.pos;
        // Payload / sig are base64 — no escapes. Keyid uses our JCS short-form
        // escapes at emit time but in practice mkit keyids are ASCII-only.
        // Reject any backslash for simplicity; we never emit them.
        while (self.pos < self.src.len and self.src[self.pos] != '"') : (self.pos += 1) {
            if (self.src[self.pos] == '\\') return error.MalformedEnvelope;
        }
        if (self.pos >= self.src.len) return error.MalformedEnvelope;
        const buf = try allocator.dupe(u8, self.src[start..self.pos]);
        self.pos += 1; // closing "
        return buf;
    }
};

// -----------------------------------------------------------------------------
// Tests
// -----------------------------------------------------------------------------

const testing = std.testing;

test "pae: DSSE v1 reference encoding" {
    // Taken straight from the DSSE spec example.
    // PAE("hello", "body") = "DSSEv1 5 hello 4 body"
    const got = try pae(testing.allocator, "hello", "body");
    defer testing.allocator.free(got);
    try testing.expectEqualStrings("DSSEv1 5 hello 4 body", got);
}

test "pae: empty payload is allowed but length is 0" {
    const got = try pae(testing.allocator, "t", "");
    defer testing.allocator.free(got);
    try testing.expectEqualStrings("DSSEv1 1 t 0 ", got);
}

test "encode: one signature, canonical shape" {
    const got = try encode(testing.allocator, .{
        .payload_type = "application/vnd.in-toto+json",
        .payload = "{}",
        .signatures = &.{.{ .keyid = "blake3:aa", .sig = "\x01\x02\x03" }},
    });
    defer testing.allocator.free(got);
    // payload "{}" = 7b 7d → base64 "e30="
    // sig 01 02 03      → base64 "AQID"
    try testing.expectEqualStrings(
        "{\"payload\":\"e30=\"," ++
            "\"payloadType\":\"application/vnd.in-toto+json\"," ++
            "\"signatures\":[{\"keyid\":\"blake3:aa\",\"sig\":\"AQID\"}]}",
        got,
    );
}

test "encode: zero signatures rejected" {
    try testing.expectError(
        error.EnvelopeNeedsAtLeastOneSignature,
        encode(testing.allocator, .{
            .payload_type = "x",
            .payload = "{}",
            .signatures = &.{},
        }),
    );
}

test "encode: empty payload_type rejected" {
    try testing.expectError(
        error.PayloadTypeEmpty,
        encode(testing.allocator, .{
            .payload_type = "",
            .payload = "{}",
            .signatures = &.{.{ .keyid = "k", .sig = "s" }},
        }),
    );
}

test "encode/decode round-trip" {
    const env: Envelope = .{
        .payload_type = PAYLOAD_TYPE_IN_TOTO,
        .payload = "{\"a\":1}",
        .signatures = &.{
            .{ .keyid = "blake3:aa", .sig = "\x10\x20\x30\x40" },
            .{ .keyid = "sigstore:https://example.com", .sig = "\xAA\xBB\xCC" },
        },
    };
    const bytes = try encode(testing.allocator, env);
    defer testing.allocator.free(bytes);

    var decoded = try decode(testing.allocator, bytes);
    defer decoded.deinit();

    try testing.expectEqualStrings(env.payload_type, decoded.payload_type);
    try testing.expectEqualStrings(env.payload, decoded.payload);
    try testing.expectEqual(@as(usize, 2), decoded.signatures.len);
    try testing.expectEqualStrings("blake3:aa", decoded.signatures[0].keyid);
    try testing.expectEqualSlices(u8, "\x10\x20\x30\x40", decoded.signatures[0].sig);
    try testing.expectEqualStrings("sigstore:https://example.com", decoded.signatures[1].keyid);
    try testing.expectEqualSlices(u8, "\xAA\xBB\xCC", decoded.signatures[1].sig);
}

test "decode: rejects malformed envelope" {
    try testing.expectError(error.MalformedEnvelope, decode(testing.allocator, "not json"));
    try testing.expectError(error.MalformedEnvelope, decode(testing.allocator, "{}"));
    // trailing garbage
    const good = try encode(testing.allocator, .{
        .payload_type = "x",
        .payload = "",
        .signatures = &.{.{ .keyid = "k", .sig = "" }},
    });
    defer testing.allocator.free(good);
    const with_trailer = try std.fmt.allocPrint(testing.allocator, "{s}trailing", .{good});
    defer testing.allocator.free(with_trailer);
    try testing.expectError(error.MalformedEnvelope, decode(testing.allocator, with_trailer));
}

test "attestationId is stable across equivalent envelopes" {
    const a = try encode(testing.allocator, .{
        .payload_type = PAYLOAD_TYPE_IN_TOTO,
        .payload = "{}",
        .signatures = &.{.{ .keyid = "k", .sig = "\x01" }},
    });
    defer testing.allocator.free(a);
    const b = try encode(testing.allocator, .{
        .payload_type = PAYLOAD_TYPE_IN_TOTO,
        .payload = "{}",
        .signatures = &.{.{ .keyid = "k", .sig = "\x01" }},
    });
    defer testing.allocator.free(b);
    try testing.expectEqual(attestationId(a), attestationId(b));
}

File: rust/crates/mkit-attest/src/envelope.rs
//! DSSE envelope — outer signed container for every mkit attestation.
//!
//! See `docs/SPEC-ATTESTATIONS.md` §2 / §4.1 and the upstream
//! <https://github.com/secure-systems-lab/dsse/blob/master/envelope.md>.
//!
//! Envelope shape (JCS-canonical key order):
//!
//! ```json
//! {
//!   "payload":     "<base64(payload_bytes)>",
//!   "payloadType": "<media-type>",
//!   "signatures":  [ { "keyid": "<...>", "sig": "<base64(sig_bytes)>" }, ... ]
//! }
//! ```
//!
//! Signed bytes are the DSSE Pre-Authentication Encoding (PAE):
//!
//! ```text
//! "DSSEv1" SP ascii(len(type)) SP type SP ascii(len(payload)) SP payload
//! ```
//!
//! mkit does not implement the hashed-PAE variant; the signer trait
//! always hands the signer the raw PAE and lets the signer decide
//! whether to pre-hash.

use base64::Engine as _;
use base64::engine::general_purpose::STANDARD as B64;

use crate::Error;
use crate::jcs::{self, Member, Value};
use mkit_core::Hash;

/// MIME type written into `payloadType` for every mkit attestation.
pub const PAYLOAD_TYPE_IN_TOTO: &str = "application/vnd.in-toto+json";

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Sig {
    pub keyid: String,
    /// Raw signature bytes (NOT base64). Encoder base64s on the way out.
    pub sig: Vec<u8>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Envelope {
    pub payload_type: String,
    /// Raw payload bytes (NOT base64). Encoder base64s on the way out.
    pub payload: Vec<u8>,
    pub signatures: Vec<Sig>,
}

impl Envelope {
    /// Build the DSSE PAE for this envelope's `(payloadType, payload)`.
    #[must_use]
    pub fn pae(&self) -> Vec<u8> {
        pae_of(&self.payload_type, &self.payload)
    }

    /// Encode this envelope as JCS-canonical JSON. Caller owns the bytes.
    ///
    /// # Errors
    /// * [`Error::EnvelopeNeedsAtLeastOneSignature`] if `signatures` is empty.
    /// * [`Error::PayloadTypeEmpty`] if `payload_type` is `""`.
    pub fn encode(&self) -> Result<String, Error> {
        encode(self)
    }

    /// BLAKE3 of the encoded envelope bytes — the att-id used as the
    /// on-disk filename stem.
    ///
    /// # Errors
    /// Same as [`Envelope::encode`].
    pub fn attestation_id(&self) -> Result<Hash, Error> {
        let bytes = self.encode()?;
        Ok(attestation_id(bytes.as_bytes()))
    }
}

/// Free-function PAE builder. Used directly by verifiers.
#[must_use]
pub fn pae_of(payload_type: &str, payload: &[u8]) -> Vec<u8> {
    let mut buf = Vec::with_capacity(8 + 8 + payload_type.len() + 8 + payload.len() + 4);
    buf.extend_from_slice(b"DSSEv1 ");
    buf.extend_from_slice(payload_type.len().to_string().as_bytes());
    buf.push(b' ');
    buf.extend_from_slice(payload_type.as_bytes());
    buf.push(b' ');
    buf.extend_from_slice(payload.len().to_string().as_bytes());
    buf.push(b' ');
    buf.extend_from_slice(payload);
    buf
}

/// Encode an envelope as JCS-canonical JSON.
///
/// # Errors
/// See [`Envelope::encode`].
pub fn encode(env: &Envelope) -> Result<String, Error> {
    if env.signatures.is_empty() {
        return Err(Error::EnvelopeNeedsAtLeastOneSignature);
    }
    if env.payload_type.is_empty() {
        return Err(Error::PayloadTypeEmpty);
    }

    let payload_b64 = B64.encode(&env.payload);

    let sig_values: Vec<Value> = env
        .signatures
        .iter()
        .map(|s| {
            Value::Object(vec![
                Member::new("keyid", Value::String(s.keyid.clone())),
                Member::new("sig", Value::String(B64.encode(&s.sig))),
            ])
        })
        .collect();

    let root = Value::Object(vec![
        Member::new("payload", Value::String(payload_b64)),
        Member::new("payloadType", Value::String(env.payload_type.clone())),
        Member::new("signatures", Value::Array(sig_values)),
    ]);

    jcs::encode(&root)
}

/// Compute the att-id for an already-serialised envelope (BLAKE3 of bytes).
#[must_use]
pub fn attestation_id(envelope_bytes: &[u8]) -> Hash {
    mkit_core::hash::hash(envelope_bytes)
}

// ---------------------------------------------------------------------------
// Decoder — strict, matches exactly the byte shape our `encode` produces.
//
// The on-disk attestations mkit creates always come from this encoder,
// so we accept only its exact byte shape and reject anything with
// non-JCS spacing. If we ever need to ingest third-party DSSE envelopes
// we can re-canonicalise via serde + this writer before storing.
// ---------------------------------------------------------------------------

/// Decode a JCS-canonical DSSE envelope produced by [`encode`].
///
/// # Errors
/// [`Error::MalformedEnvelope`] if the input does not match the exact
/// byte shape this crate emits, or if the embedded base64 payload /
/// signatures fail to decode.
pub fn decode(bytes: &[u8]) -> Result<Envelope, Error> {
    let mut p = Parser { src: bytes, pos: 0 };
    p.expect(b"{\"payload\":")?;
    let payload_b64 = p.take_string()?;
    p.expect(b",\"payloadType\":")?;
    let payload_type = p.take_string()?;
    p.expect(b",\"signatures\":[")?;

    let mut sigs: Vec<Sig> = Vec::new();
    if !p.peek(b']') {
        loop {
            p.expect(b"{\"keyid\":")?;
            let keyid = p.take_string()?;
            p.expect(b",\"sig\":")?;
            let sig_b64 = p.take_string()?;
            p.expect(b"}")?;

            let sig_bytes = B64
                .decode(sig_b64.as_bytes())
                .map_err(|_| Error::MalformedEnvelope)?;
            sigs.push(Sig {
                keyid,
                sig: sig_bytes,
            });

            if p.peek(b',') {
                p.pos += 1;
                continue;
            }
            break;
        }
    }
    p.expect(b"]}")?;
    if p.pos != bytes.len() {
        return Err(Error::MalformedEnvelope);
    }

    let payload = B64
        .decode(payload_b64.as_bytes())
        .map_err(|_| Error::MalformedEnvelope)?;

    Ok(Envelope {
        payload_type,
        payload,
        signatures: sigs,
    })
}

struct Parser<'a> {
    src: &'a [u8],
    pos: usize,
}

impl Parser<'_> {
    fn expect(&mut self, s: &[u8]) -> Result<(), Error> {
        let end = self
            .pos
            .checked_add(s.len())
            .ok_or(Error::MalformedEnvelope)?;
        if end > self.src.len() {
            return Err(Error::MalformedEnvelope);
        }
        if &self.src[self.pos..end] != s {
            return Err(Error::MalformedEnvelope);
        }
        self.pos = end;
        Ok(())
    }

    fn peek(&self, c: u8) -> bool {
        self.pos < self.src.len() && self.src[self.pos] == c
    }

    /// Take a JSON string, returning an owned `String`. Backslash
    /// escapes are NOT supported (mkit never emits them inside an
    /// envelope key/value position — payload/sig are base64; keyids are
    /// ASCII URLs / hex).
    fn take_string(&mut self) -> Result<String, Error> {
        if self.pos >= self.src.len() || self.src[self.pos] != b'"' {
            return Err(Error::MalformedEnvelope);
        }
        self.pos += 1;
        let start = self.pos;
        while self.pos < self.src.len() && self.src[self.pos] != b'"' {
            if self.src[self.pos] == b'\\' {
                return Err(Error::MalformedEnvelope);
            }
            self.pos += 1;
        }
        if self.pos >= self.src.len() {
            return Err(Error::MalformedEnvelope);
        }
        let s = core::str::from_utf8(&self.src[start..self.pos])
            .map_err(|_| Error::MalformedEnvelope)?
            .to_owned();
        self.pos += 1; // closing "
        Ok(s)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn pae_dsse_v1_reference_encoding() {
        // From the DSSE spec: PAE("hello", "body") = "DSSEv1 5 hello 4 body"
        let got = pae_of("hello", b"body");
        assert_eq!(got, b"DSSEv1 5 hello 4 body");
    }

    #[test]
    fn pae_empty_payload() {
        let got = pae_of("t", b"");
        assert_eq!(got, b"DSSEv1 1 t 0 ");
    }

    #[test]
    fn encode_one_signature_canonical_shape() {
        let env = Envelope {
            payload_type: PAYLOAD_TYPE_IN_TOTO.into(),
            payload: b"{}".to_vec(),
            signatures: vec![Sig {
                keyid: "blake3:aa".into(),
                sig: vec![0x01, 0x02, 0x03],
            }],
        };
        let got = env.encode().unwrap();
        // payload "{}" = 7b 7d → base64 "e30="
        // sig 01 02 03      → base64 "AQID"
        assert_eq!(
            got,
            "{\"payload\":\"e30=\",\
             \"payloadType\":\"application/vnd.in-toto+json\",\
             \"signatures\":[{\"keyid\":\"blake3:aa\",\"sig\":\"AQID\"}]}"
        );
    }

    #[test]
    fn encode_zero_signatures_rejected() {
        let env = Envelope {
            payload_type: "x".into(),
            payload: b"{}".to_vec(),
            signatures: vec![],
        };
        assert!(matches!(
            env.encode(),
            Err(Error::EnvelopeNeedsAtLeastOneSignature)
        ));
    }

    #[test]
    fn encode_empty_payload_type_rejected() {
        let env = Envelope {
            payload_type: String::new(),
            payload: b"{}".to_vec(),
            signatures: vec![Sig {
                keyid: "k".into(),
                sig: vec![1],
            }],
        };
        assert!(matches!(env.encode(), Err(Error::PayloadTypeEmpty)));
    }

    #[test]
    fn encode_decode_roundtrip() {
        let env = Envelope {
            payload_type: PAYLOAD_TYPE_IN_TOTO.into(),
            payload: b"{\"a\":1}".to_vec(),
            signatures: vec![
                Sig {
                    keyid: "blake3:aa".into(),
                    sig: vec![0x10, 0x20, 0x30, 0x40],
                },
                Sig {
                    keyid: "sigstore:https://example.com".into(),
                    sig: vec![0xAA, 0xBB, 0xCC],
                },
            ],
        };
        let bytes = env.encode().unwrap();
        let dec = decode(bytes.as_bytes()).unwrap();
        assert_eq!(dec, env);
    }

    #[test]
    fn decode_rejects_malformed() {
        assert!(matches!(decode(b"not json"), Err(Error::MalformedEnvelope)));
        assert!(matches!(decode(b"{}"), Err(Error::MalformedEnvelope)));

        // Trailing garbage.
        let env = Envelope {
            payload_type: "x".into(),
            payload: vec![],
            signatures: vec![Sig {
                keyid: "k".into(),
                sig: vec![],
            }],
        };
        let mut bad = env.encode().unwrap();
        bad.push_str("trailing");
        assert!(matches!(
            decode(bad.as_bytes()),
            Err(Error::MalformedEnvelope)
        ));
    }

    #[test]
    fn attestation_id_stable_across_equivalent_envelopes() {
        let a = Envelope {
            payload_type: PAYLOAD_TYPE_IN_TOTO.into(),
            payload: b"{}".to_vec(),
            signatures: vec![Sig {
                keyid: "k".into(),
                sig: vec![1],
            }],
        };
        let b = a.clone();
        assert_eq!(a.attestation_id().unwrap(), b.attestation_id().unwrap());
    }
}
